### PR TITLE
Add cmake scripts for iwyu targets.

### PR DESCRIPTION
**Pull request by nafur**
This PR adds some utility targets that simplify the usage of iwyu (include-what-you-use) on our code base.

### CODE REVIEW

File: cmake/IWYU.cmake [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+if(IWYU_PATH)
+    # iwyu is available
+
+    # add a target to inspect number of times headers are included.
+    add_custom_target(iwyu-list-includes
+        COMMAND grep -hor "\"#include .*\""  ${PROJECT_SOURCE_DIR}/src | sort | uniq -c | sort -n
+    )
+
+    # find the standard library directory
+    file(GLOB_RECURSE LLVM_INCLUDE /usr/lib*/llvm*/**/stddef.h)
```
Comment (mpreiner) [inline cmake/IWYU.cmake:12]: I'd not use GLOB_RECURSE here. Finding a directory should be possible via `find_path`.

File: cmake/IWYU.cmake [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    if(LLVM_INCLUDE)
+        list(GET LLVM_INCLUDE 0 LLVM_INCLUDE)
+        get_filename_component(LLVM_INCLUDE "${LLVM_INCLUDE}" DIRECTORY)
+        set(LLVM_INCLUDE "-I${LLVM_INCLUDE}")
+    endif()
+
+    # add a target to run iwyu on all files (within the compilation database)
+    add_custom_target(iwyu-all
+        COMMAND ${IWYU_PATH} -o clang -p . -- -Xiwyu --cxx17ns ${LLVM_INCLUDE}
+    )
```
Comment (mpreiner) [inline cmake/IWYU.cmake:22]: What happens if `LLVM_INCLUDE` is empty?

File: cmake/IWYU.cmake (hunk 1 above)
Comment (nafur) [reply cmake/IWYU.cmake:12]: I tried and failed. While `find_path` has some support for wildcards, it is very limited and does not cover this case. Even if we turn `lib*` into `lib` (and `lib64`), I still could not get it to work.

File: cmake/IWYU.cmake (hunk 2 above)
Comment (nafur) [reply cmake/IWYU.cmake:22]: On my machine, it complains about not finding `stddef.h`. I'm not really sure why that is, though.

File: cmake/IWYU.cmake [hunk 3]
```diff
@@ -0,0 +1,36 @@
+find_program(IWYU_PATH iwyu_tool)
```
Comment (mpreiner) [inline cmake/IWYU.cmake:1]:
```suggestion
find_program(IWYU_PATH NAMES iwyu_tool iwyu-tool)
```

On Arch Linux the tool is called `iwyu-tool`

File: cmake/IWYU.cmake [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+if(IWYU_PATH)
+    # iwyu is available
```
Comment (mpreiner) [inline cmake/IWYU.cmake:4]:
```suggestion
    # iwyu is available
    message(STATUS "Found IWYU: ${IWYU_PATH}")
```

File: cmake/IWYU.cmake (hunk 1 above)
Comment (mpreiner) [reply cmake/IWYU.cmake:12]: The IWYU documentation suggests to use `clang -print-resource-dir` to find the correct clang directory (https://github.com/include-what-you-use/include-what-you-use#how-to-install).

File: cmake/IWYU.cmake [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    file(GLOB subdirs ${PROJECT_SOURCE_DIR}/src/*)
+    foreach(dir ${subdirs})
+        # add a target for every subdirectory
+        file(GLOB_RECURSE source_files
+            ${dir}/*.cpp
+            ${dir}/*.cc
+        )
+        get_filename_component(dirname ${dir} NAME)
+        add_custom_target(iwyu-${dirname}
+            COMMAND ${IWYU_PATH} -o clang -j4 -p . ${source_files} -- -Xiwyu --cxx17ns ${LLVM_INCLUDE}
```
Comment (mpreiner) [inline cmake/IWYU.cmake:33]: Not sure if `-j4` should be hard coded. If we want to use all available threads we can use `CTEST_NTHREADS`, which is defined in the top-level `CMakeLists.txt`

Comment (nafur) [reply cmake/IWYU.cmake]: So what I would like to achieve is that we can do `make -j8 iwyu-all`. I've not found a way (yet), and simply removed the `-j4` for now.

File: cmake/IWYU.cmake [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    # iwyu is available
+    message(STATUS "Found IWYU: ${IWYU_PATH}")
+
+    # add a target to inspect number of times headers are included.
+    add_custom_target(iwyu-list-includes
+        COMMAND grep -hor "\"#include .*\""  ${PROJECT_SOURCE_DIR}/src | sort | uniq -c | sort -n
+    )
+
+    # find the standard library directory
+    execute_process(COMMAND clang -print-resource-dir
```
Comment (mpreiner) [inline cmake/IWYU.cmake:13]: Might be a good idea to first check via `find_program(CLANG_PATH clang)` if clang is installed on the system and use `${CLANG_PATH}` in this command.

**Review by mpreiner — APPROVED** [review]
LGTM, thanks for the changes!

**Pull request merged**